File: android/src/main/java/com/posthog/posthog_flutter/PosthogFlutterPlugin.java
package com.posthog.posthog_flutter;

import androidx.annotation.NonNull;

import android.app.Activity;
import android.content.Context;
import android.content.pm.ApplicationInfo;
import android.content.pm.PackageManager;
import android.content.SharedPreferences;
import android.os.Bundle;
import android.util.Log;

import com.posthog.android.PostHog;
import com.posthog.android.PostHogContext;
import com.posthog.android.Properties;
import com.posthog.android.Options;
import com.posthog.android.Middleware;
import com.posthog.android.payloads.BasePayload;
import static com.posthog.android.PostHog.LogLevel;

import java.util.LinkedHashMap;
import java.util.HashMap;
import java.util.Map;

import io.flutter.plugin.common.BinaryMessenger;
import io.flutter.plugin.common.MethodCall;
import io.flutter.plugin.common.MethodChannel;
import io.flutter.plugin.common.MethodChannel.MethodCallHandler;
import io.flutter.plugin.common.MethodChannel.Result;
import io.flutter.plugin.common.PluginRegistry;
import io.flutter.embedding.engine.plugins.FlutterPlugin;

/** PosthogFlutterPlugin */
public class PosthogFlutterPlugin implements MethodCallHandler, FlutterPlugin {
  private Context applicationContext;
  private MethodChannel methodChannel;

  static HashMap<String, Object> appendToContextMiddleware;

  /** Plugin registration. */
  public static void registerWith(PluginRegistry.Registrar registrar) {
    final PosthogFlutterPlugin instance = new PosthogFlutterPlugin();
    instance.setupChannels(registrar.context(), registrar.messenger());
  }

  @Override
  public void onAttachedToEngine(@NonNull FlutterPluginBinding binding) {
    setupChannels(binding.getApplicationContext(), binding.getBinaryMessenger());
  }

  private void setupChannels(Context applicationContext, BinaryMessenger messenger) {
    try {
      methodChannel = new MethodChannel(messenger, "posthogflutter");
      this.applicationContext = applicationContext;

      ApplicationInfo ai = applicationContext.getPackageManager()
        .getApplicationInfo(applicationContext.getPackageName(), PackageManager.GET_META_DATA);

      Bundle bundle = ai.metaData;

      String writeKey = bundle.getString("com.posthog.posthog.API_KEY");
      String posthogHost = bundle.getString("com.posthog.posthog.POSTHOG_HOST");
      Boolean captureApplicationLifecycleEvents = bundle.getBoolean("com.posthog.posthog.TRACK_APPLICATION_LIFECYCLE_EVENTS");
      Boolean debug = bundle.getBoolean("com.posthog.posthog.DEBUG", false);

      PostHog.Builder posthogBuilder = new PostHog.Builder(applicationContext, writeKey, posthogHost);
      if (captureApplicationLifecycleEvents) {
        // Enable this to record certain application events automatically
        posthogBuilder.captureApplicationLifecycleEvents();
      }

      if (debug) {
        posthogBuilder.logLevel(LogLevel.DEBUG);
      }

      // Here we build a middleware that just appends data to the current context
      // using the [deepMerge] strategy.
      posthogBuilder.middleware(
        new Middleware() {
          @Override
          public void intercept(Chain chain) {
            try {
              if (appendToContextMiddleware == null) {
                chain.proceed(chain.payload());
                return;
              }

              BasePayload payload = chain.payload();
              BasePayload newPayload = payload.toBuilder()
                .context(appendToContextMiddleware)
                .build();

              chain.proceed(newPayload);
            } catch (Exception e) {
              Log.e("PosthogFlutter", e.getMessage());
              chain.proceed(chain.payload());
            }
          }
        }
      );

      // Set the initialized instance as globally accessible.
      // It may throw an exception if we are trying to re-register a singleton PostHog instance.
      // This state may happen after the app is popped (back button until the app closes)
      // and opened again from the TaskManager.
      try {
        PostHog.setSingletonInstance(posthogBuilder.build());
      } catch (IllegalStateException e) {
        Log.w("PosthogFlutter", e.getMessage());
      }
      // register the channel to receive calls
      methodChannel.setMethodCallHandler(this);
    } catch (Exception e) {
      Log.e("PosthogFlutter", e.getMessage());
    }
  }

  @Override
  public void onDetachedFromEngine(FlutterPluginBinding binding) { }

  @Override
  public void onMethodCall(MethodCall call, Result result) {
    if(call.method.equals("identify")) {
      this.identify(call, result);
    } else if (call.method.equals("capture")) {
      this.capture(call, result);
    } else if (call.method.equals("screen")) {
      this.screen(call, result);
    } else if (call.method.equals("alias")) {
      this.alias(call, result);
    } else if (call.method.equals("getAnonymousId")) {
      this.anonymousId(result);
    } else if (call.method.equals("reset")) {
      this.reset(result);
    } else if (call.method.equals("setContext")) {
      this.setContext(call, result);
    } else if (call.method.equals("disable")) {
      this.disable(call, result);
    } else if (call.method.equals("enable")) {
      this.enable(call, result);
    } else if (call.method.equals("isFeatureEnabled")) {
      this.isFeatureEnabled(call, result);
    } else if (call.method.equals("reloadFeatureFlags")) {
      this.reloadFeatureFlags(call, result);
    } else if (call.method.equals("group")) {
      this.reloadFeatureFlags(call, result);
    } else {
      result.notImplemented();
    }
  }

  private Properties hashMapToProperties(HashMap<String, Object> propertiesData) {
    Properties properties = new Properties();

    for(Map.Entry<String, Object> property : propertiesData.entrySet()) {
      String key = property.getKey();
      Object value = property.getValue();
      properties.putValue(key, value);
    }

    return properties;
  }

  private void identify(MethodCall call, Result result) {
    try {
      String userId = call.argument("userId");
      HashMap<String, Object> propertiesData = call.argument("properties");
      HashMap<String, Object> optionsData = call.argument("options");
      Properties properties = this.hashMapToProperties(propertiesData);
      Options options = this.buildOptions(optionsData);
      PostHog.with(this.applicationContext).identify(userId, properties, options);

      result.success(true);
    } catch (Exception e) {
      result.error("PosthogFlutterException", e.getLocalizedMessage(), null);
    }
  }

  private void capture(MethodCall call, Result result) {
    try {
      String eventName = call.argument("eventName");
      HashMap<String, Object> propertiesData = call.argument("properties");
      HashMap<String, Object> optionsData = call.argument("options");
      Properties properties = this.hashMapToProperties(propertiesData);
      Options options = this.buildOptions(optionsData);

      PostHog.with(this.applicationContext).capture(eventName, properties, options);
      result.success(true);
    } catch (Exception e) {
      result.error("PosthogFlutterException", e.getLocalizedMessage(), null);
    }
  }

  private void screen(MethodCall call, Result result) {
    try {
      String screenName = call.argument("screenName");
      HashMap<String, Object> propertiesData = call.argument("properties");
      HashMap<String, Object> optionsData = call.argument("options");
      Properties properties = this.hashMapToProperties(propertiesData);
      Options options = this.buildOptions(optionsData);

      PostHog.with(this.applicationContext).screen(screenName, properties, options);
      result.success(true);
    } catch (Exception e) {
      result.error("PosthogFlutterException", e.getLocalizedMessage(), null);
    }
  }

  private void alias(MethodCall call, Result result) {
    try {
      String alias = call.argument("alias");
      HashMap<String, Object> optionsData = call.argument("options");
      Options options = this.buildOptions(optionsData);
      PostHog.with(this.applicationContext).alias(alias, options);
      result.success(true);
    } catch (Exception e) {
      result.error("PosthogFlutterException", e.getLocalizedMessage(), null);
    }
  }

  private void anonymousId(Result result) {
    try {
      String anonymousId = PostHog.with(this.applicationContext).getAnonymousId();
      result.success(anonymousId);
    } catch (Exception e) {
      result.error("PosthogFlutterException", e.getLocalizedMessage(), null);
    }
  }

  private void reset(Result result) {
    try {
      PostHog.with(this.applicationContext).reset();
      result.success(true);
    } catch (Exception e) {
      result.error("PosthogFlutterException", e.getLocalizedMessage(), null);
    }
  }

  private void setContext(MethodCall call, Result result) {
    try {
      this.appendToContextMiddleware = call.argument("context");
      result.success(true);
    } catch (Exception e) {
      result.error("PosthogFlutterException", e.getLocalizedMessage(), null);
    }
  }

  // There is no enable method at this time for PostHog on Android.
  // Instead, we use optOut as a proxy to achieve the same result.
  private void enable(MethodCall call, Result result) {
    try {
      PostHog.with(this.applicationContext).optOut(false);
      result.success(true);
    } catch (Exception e) {
      result.error("PosthogFlutterException", e.getLocalizedMessage(), null);
    }
  }

  // There is no disable method at this time for PostHog on Android.
  // Instead, we use optOut as a proxy to achieve the same result.
  private void disable(MethodCall call, Result result) {
    try {
      PostHog.with(this.applicationContext).optOut(true);
      result.success(true);
    } catch (Exception e) {
      result.error("PosthogFlutterException", e.getLocalizedMessage(), null);
    }
  }

  private void isFeatureEnabled(MethodCall call, Result result) {
    try {
      String key = call.argument("key");
      boolean isEnabled = PostHog.with(this.applicationContext).isFeatureEnabled(key);
      result.success(isEnabled);
    } catch (Exception e) {
      result.error("PosthogFlutterException", e.getLocalizedMessage(), null);
    }
  }

  private void reloadFeatureFlags(MethodCall call, Result result) {
    try {
      PostHog.with(this.applicationContext).reloadFeatureFlags();
      result.success(true);
    } catch (Exception e) {
      result.error("PosthogFlutterException", e.getLocalizedMessage(), null);
    }
  }

  private void group(MethodCall call, Result result) {
    try {
      String groupType = call.argument("groupType");
      String groupKey = call.argument("groupKey");
      HashMap<String, Object> propertiesData = call.argument("groupProperties");
      Properties properties = this.hashMapToProperties(propertiesData);

      PostHog.with(this.applicationContext).group(groupType, groupKey, properties);
      result.success(true);
    } catch (Exception e) {
      result.error("PosthogFlutterException", e.getLocalizedMessage(), null);
    }
  }
  /**
   * Enables / disables / sets custom integration properties so Posthog can properly
   * interact with 3rd parties, such as Amplitude.
   * @see https://posthog.com/docs/connections/sources/catalog/libraries/mobile/android/#selecting-destinations
   * @see https://github.com/posthogio/posthog-android/blob/master/posthog/src/main/java/com/posthog.android/Options.java
   */
  @SuppressWarnings("unchecked")
  private Options buildOptions(HashMap<String, Object> optionsData) {
    Options options = new Options();
    return options;
  }

  // Merges [newMap] into [original], *not* preserving [original]
  // keys (deep) in case of conflicts.
  private static Map deepMerge(Map original, Map newMap) {
    for (Object key : newMap.keySet()) {
      if (newMap.get(key) instanceof Map && original.get(key) instanceof Map) {
        Map originalChild = (Map) original.get(key);
        Map newChild = (Map) newMap.get(key);
        original.put(key, deepMerge(originalChild, newChild));
      } else {
        original.put(key, newMap.get(key));
      }
    }
    return original;
  }
}

File: android/src/main/kotlin/com/posthog/posthog_flutter/PosthogFlutterPlugin.kt
package com.posthog.posthog_flutter

import android.content.Context
import android.graphics.Bitmap
import android.graphics.BitmapFactory
import android.util.Log
import com.posthog.PostHog
import com.posthog.android.PostHogAndroid
import com.posthog.android.PostHogAndroidConfig
import com.posthog.internal.replay.RRFullSnapshotEvent
import com.posthog.internal.replay.RRStyle
import com.posthog.internal.replay.RRWireframe
import com.posthog.internal.replay.capture
import io.flutter.embedding.engine.plugins.FlutterPlugin
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel
import io.flutter.plugin.common.MethodChannel.MethodCallHandler
import io.flutter.plugin.common.MethodChannel.Result
import java.io.ByteArrayOutputStream

/** PosthogFlutterPlugin */
class PosthogFlutterPlugin : FlutterPlugin, MethodCallHandler {
    /// The MethodChannel that will the communication between Flutter and native Android
    ///
    /// This local reference serves to register the plugin with the Flutter Engine and unregister it
    /// when the Flutter Engine is detached from the Activity
    private lateinit var channel: MethodChannel

    private lateinit var context: Context

    override fun onAttachedToEngine(flutterPluginBinding: FlutterPlugin.FlutterPluginBinding) {
        channel = MethodChannel(flutterPluginBinding.binaryMessenger, "posthog_flutter")

        context = flutterPluginBinding.applicationContext

        channel.setMethodCallHandler(this)
    }

    override fun onMethodCall(call: MethodCall, result: Result) {

        when (call.method) {

            "identify" -> {
                identify(call, result)
            }

            "capture" -> {
                capture(call, result)
            }

            "screen" -> {
                screen(call, result)
            }

            "alias" -> {
                alias(call, result)
            }

            "distinctId" -> {
                distinctId(result)
            }

            "reset" -> {
                reset(result)
            }

            "disable" -> {
                disable(result)
            }

            "enable" -> {
                enable(result)
            }

            "isFeatureEnabled" -> {
                isFeatureEnabled(call, result)
            }

            "reloadFeatureFlags" -> {
                reloadFeatureFlags(result)
            }

            "group" -> {
                group(call, result)
            }

            "getFeatureFlag" -> {
                getFeatureFlag(call, result)
            }

            "getFeatureFlagPayload" -> {
                getFeatureFlagPayload(call, result)
            }

            "register" -> {
                register(call, result)
            }

            "unregister" -> {
                unregister(call, result)
            }

            "debug" -> {
                debug(call, result)
            }

            "flush" -> {
                flush(result)
            }

            "initNativeSdk" -> {
                val configMap = call.arguments as? Map<String, Any>
                println()
                if (configMap != null) {
                    initPlugin(configMap)
                    result.success(null)
                } else {
                    result.error("INVALID_ARGUMENT", "Config map is null or invalid", null)
                }
            }

            "sendReplayScreenshot" -> {
                sendReplayScreenshot(call, result)
            }

            else -> {
                result.notImplemented()
            }
        }

    }

    private fun initPlugin(configMap: Map<String, Any>) {
        val apiKey = configMap["apiKey"] as? String ?: ""
        val options = configMap["options"] as? Map<String, Any> ?: emptyMap()

        val captureNativeAppLifecycleEvents =
            options["captureNativeAppLifecycleEvents"] as? Boolean ?: false
        val enableSessionReplay = options["enableSessionReplay"] as? Boolean ?: false
        val sessionReplayConfigMap =
            options["sessionReplayConfig"] as? Map<String, Any> ?: emptyMap()

        val maskAllTextInputs = sessionReplayConfigMap["maskAllTextInputs"] as? Boolean ?: true
        val maskAllImages = sessionReplayConfigMap["maskAllImages"] as? Boolean ?: true
        val captureLog = sessionReplayConfigMap["captureLog"] as? Boolean ?: true
        val debouncerDelayMs =
            (sessionReplayConfigMap["androidDebouncerDelayMs"] as? Int ?: 500).toLong()

        val config = PostHogAndroidConfig(apiKey).apply {
            debug = false
            captureDeepLinks = false
            captureApplicationLifecycleEvents = captureNativeAppLifecycleEvents
            captureScreenViews = false
            sessionReplay = enableSessionReplay
            sessionReplayConfig.screenshot = true
            sessionReplayConfig.captureLogcat = captureLog
            sessionReplayConfig.debouncerDelayMs = debouncerDelayMs
            sessionReplayConfig.maskAllImages = maskAllImages
            sessionReplayConfig.maskAllTextInputs = maskAllTextInputs
        }

        PostHogAndroid.setup(context, config)
    }

    /*
    * TEMPORARY FUNCTION FOR TESTING PURPOSES
    * This function sends a screenshot to PostHog.
    * It should be removed or refactored in the other version.
    */
    private fun sendReplayScreenshot(call: MethodCall, result: Result) {
        val imageBytes = call.argument<ByteArray>("imageBytes")
        if (imageBytes != null) {
            val bitmap = BitmapFactory.decodeByteArray(imageBytes, 0, imageBytes.size)

            val base64String = bitmapToBase64(bitmap)

            val wireframe = RRWireframe(
                id = System.identityHashCode(bitmap),
                x = 0,
                y = 0,
                width = bitmap.width,
                height = bitmap.height,
                type = "screenshot",
                base64 = base64String,
                style = RRStyle()
            )

            val snapshotEvent = RRFullSnapshotEvent(
                listOf(wireframe),
                initialOffsetTop = 0,
                initialOffsetLeft = 0,
                timestamp = System.currentTimeMillis()
            )

            listOf(snapshotEvent).capture()
            if (base64String != null) {
                Log.d("PostHog", base64String)
            }
        } else {
            result.error("INVALID_ARGUMENT", "Image bytes are null", null)
        }
    }

    private fun bitmapToBase64(bitmap: Bitmap): String? {
        ByteArrayOutputStream().use { byteArrayOutputStream ->
            bitmap.compress(
                Bitmap.CompressFormat.JPEG,
                30,
                byteArrayOutputStream
            )
            val byteArray = byteArrayOutputStream.toByteArray()
            return android.util.Base64.encodeToString(byteArray, android.util.Base64.NO_WRAP)
        }
    }

    override fun onDetachedFromEngine(binding: FlutterPlugin.FlutterPluginBinding) {
        channel.setMethodCallHandler(null)
    }

    private fun getFeatureFlag(call: MethodCall, result: Result) {
        try {
            val featureFlagKey: String = call.argument("key")!!
            val flag = PostHog.getFeatureFlag(featureFlagKey)
            result.success(flag)
        } catch (e: Throwable) {
            result.error("PosthogFlutterException", e.localizedMessage, null)
        }
    }

    private fun getFeatureFlagPayload(call: MethodCall, result: Result) {
        try {
            val featureFlagKey: String = call.argument("key")!!
            val flag = PostHog.getFeatureFlagPayload(featureFlagKey)
            result.success(flag)
        } catch (e: Throwable) {
            result.error("PosthogFlutterException", e.localizedMessage, null)
        }
    }

    private fun identify(call: MethodCall, result: Result) {
        try {
            val userId: String = call.argument("userId")!!
            val userProperties: Map<String, Any>? = call.argument("userProperties")
            val userPropertiesSetOnce: Map<String, Any>? = call.argument("userPropertiesSetOnce")
            PostHog.identify(userId, userProperties, userPropertiesSetOnce)
            result.success(null)
        } catch (e: Throwable) {
            result.error("PosthogFlutterException", e.localizedMessage, null)
        }
    }

    private fun capture(call: MethodCall, result: Result) {
        try {
            val eventName: String = call.argument("eventName")!!
            val properties: Map<String, Any>? = call.argument("properties")
            PostHog.capture(eventName, properties = properties)
            result.success(null)
        } catch (e: Throwable) {
            result.error("PosthogFlutterException", e.localizedMessage, null)
        }
    }

    private fun screen(call: MethodCall, result: Result) {
        try {
            val screenName: String = call.argument("screenName")!!
            val properties: Map<String, Any>? = call.argument("properties")
            PostHog.screen(screenName, properties)
            result.success(null)
        } catch (e: Throwable) {
            result.error("PosthogFlutterException", e.localizedMessage, null)
        }
    }

    private fun alias(call: MethodCall, result: Result) {
        try {
            val alias: String = call.argument("alias")!!
            PostHog.alias(alias)
            result.success(null)
        } catch (e: Throwable) {
            result.error("PosthogFlutterException", e.localizedMessage, null)
        }
    }

    private fun distinctId(result: Result) {
        try {
            val distinctId: String = PostHog.distinctId()
            result.success(distinctId)
        } catch (e: Throwable) {
            result.error("PosthogFlutterException", e.localizedMessage, null)
        }
    }

    private fun reset(result: Result) {
        try {
            PostHog.reset()
            result.success(null)
        } catch (e: Throwable) {
            result.error("PosthogFlutterException", e.localizedMessage, null)
        }
    }

    private fun enable(result: Result) {
        try {
            PostHog.optIn()
            result.success(null)
        } catch (e: Throwable) {
            result.error("PosthogFlutterException", e.localizedMessage, null)
        }
    }

    private fun debug(call: MethodCall, result: Result) {
        try {
            val debug: Boolean = call.argument("debug")!!
            PostHog.debug(debug)
            result.success(null)
        } catch (e: Throwable) {
            result.error("PosthogFlutterException", e.localizedMessage, null)
        }
    }

    private fun disable(result: Result) {
        try {
            PostHog.optOut()
            result.success(null)
        } catch (e: Throwable) {
            result.error("PosthogFlutterException", e.localizedMessage, null)
        }
    }

    private fun isFeatureEnabled(call: MethodCall, result: Result) {
        try {
            val key: String = call.argument("key")!!
            val isEnabled = PostHog.isFeatureEnabled(key)
            result.success(isEnabled)
        } catch (e: Throwable) {
            result.error("PosthogFlutterException", e.localizedMessage, null)
        }
    }

    private fun reloadFeatureFlags(result: Result) {
        try {
            PostHog.reloadFeatureFlags()
            result.success(null)
        } catch (e: Throwable) {
            result.error("PosthogFlutterException", e.localizedMessage, null)
        }
    }

    private fun group(call: MethodCall, result: Result) {
        try {
            val groupType: String = call.argument("groupType")!!
            val groupKey: String = call.argument("groupKey")!!
            val groupProperties: Map<String, Any>? = call.argument("groupProperties")
            PostHog.group(groupType, groupKey, groupProperties)
            result.success(null)
        } catch (e: Throwable) {
            result.error("PosthogFlutterException", e.localizedMessage, null)
        }
    }

    private fun register(call: MethodCall, result: Result) {
        try {
            val key: String = call.argument("key")!!
            val value: Any = call.argument("value")!!
            PostHog.register(key, value)
            result.success(null)
        } catch (e: Throwable) {
            result.error("PosthogFlutterException", e.localizedMessage, null)
        }
    }

    private fun unregister(call: MethodCall, result: Result) {
        try {
            val key: String = call.argument("key")!!
            PostHog.unregister(key)
            result.success(null)
        } catch (e: Throwable) {
            result.error("PosthogFlutterException", e.localizedMessage, null)
        }
    }

    private fun flush(result: Result) {
        try {
            PostHog.flush()
            result.success(null)
        } catch (e: Throwable) {
            result.error("PosthogFlutterException", e.localizedMessage, null)
        }
    }
}
